File: web/src/decode/ctc.ts
/**
 * CTC greedy デコード (フレーム位置付き).
 *
 * 移植元: src/infer/engine.py の ctc_greedy_decode_with_frames
 *
 * フレーム位置を残すのは、スライディングウィンドウで重複区間を取り除き、
 * 確定/暫定の境界を絶対時刻で判定するため。
 */
import { BLANK_TOKEN_ID, WORD_BREAK_TOKEN_ID } from '../generated/tokens'

/** 1 つのデコード済みトークン。frameStart〜frameEnd は inclusive。 */
export interface FrameToken {
  tokenId: number
  confidence: number
  frameStart: number
  frameEnd: number
}

/**
 * log_softmax 済みの `(nFrames, vocabSize)` を平坦化した配列からトークン列を作る。
 *
 * @param logProbs 長さ nFrames * vocabSize の平坦配列 (行優先)
 * @param nFrames 時間フレーム数
 * @param vocabSize 語彙サイズ
 * @param blankId CTC blank の token id
 */
export function ctcGreedyDecodeWithFrames(
  logProbs: Float32Array,
  nFrames: number,
  vocabSize: number,
  blankId: number = BLANK_TOKEN_ID,
  wordBreakBias: number = 0,
): FrameToken[] {
  // 長さが足りないと `logProbs[base + v]` が undefined になり、比較が全部 false に
  // なって「全フレームが blank」と解釈される。例外ではなく空配列が返るため、
  // 呼び出し側からは「無音だった」と区別が付かない (Python 版は ValueError)。
  if (nFrames < 0 || vocabSize <= 0) {
    throw new Error(`invalid shape: nFrames=${nFrames}, vocabSize=${vocabSize}`)
  }
  if (logProbs.length < nFrames * vocabSize) {
    throw new Error(
      `logProbs length ${logProbs.length} < nFrames * vocabSize (${nFrames} * ${vocabSize} = ${nFrames * vocabSize})`,
    )
  }
  const out: FrameToken[] = []
  let prev = -1
  let runStart = 0
  let runMax = 0

  for (let t = 0; t < nFrames; t++) {
    const base = t * vocabSize
    let best = 0
    let bestLogProb = logProbs[base]
    for (let v = 1; v < vocabSize; v++) {
      // WORD_BREAK にだけ下駄を履かせる (0 なら何もしない = 既定)。
      // 語間が出やすくなるほか、モデルが余計に出すプロサイン ([SK] 等) を
      // 押さえる効果が実測で確認されている (docs/word_break_bias.md)。
      const value =
        v === WORD_BREAK_TOKEN_ID ? logProbs[base + v] + wordBreakBias : logProbs[base + v]
      if (value > bestLogProb) {
        bestLogProb = value
        best = v
      }
    }
    const conf = Math.exp(bestLogProb)

    if (best === prev) {
      if (best !== blankId && conf > runMax) runMax = conf
      continue
    }
    // 切り替わり: 直前のランを確定する
    if (prev !== -1 && prev !== blankId) {
      out.push({ tokenId: prev, confidence: runMax, frameStart: runStart, frameEnd: t - 1 })
    }
    prev = best
    runStart = t
    runMax = conf
  }

  if (prev !== -1 && prev !== blankId) {
    out.push({ tokenId: prev, confidence: runMax, frameStart: runStart, frameEnd: nFrames - 1 })
  }
  return out
}

File: src/infer/ctc.py
"""CTC greedy デコード (フレーム位置付き) — **numpy だけで動く**.

``engine.py`` から切り出した。切り出した理由は配布物の大きさである:
PyTorch は 2.8 GB あり、インストーラの 7 割を占める。デコードに必要なのは
``exp`` と ``argmax`` だけなので、ここを numpy にしておけば
ONNX 推論の経路が torch を一切読み込まずに済む。

**このモジュールは torch を import してはいけない。**
``tests/test_no_torch_import.py`` が歯止めになっている。
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from src.tokens.morse_tokens import BLANK_TOKEN_ID


@dataclass(frozen=True)
class FrameToken:
    """1 つのデコード済みトークン (フレーム位置付き)."""

    token_id: int
    confidence: float
    frame_start: int    # この token を出力した最初のフレーム
    frame_end: int      # 最後のフレーム (inclusive)


def ctc_greedy_decode_frames(
    log_probs: np.ndarray, blank_id: int = BLANK_TOKEN_ID
) -> list[list[FrameToken]]:
    """``(B, T, V)`` の log-softmax から ``FrameToken`` 列を返す.

    各 ``FrameToken`` は ``[frame_start, frame_end]`` の連続フレームから
    出力された 1 つのトークン。blank と重複は collapse 済み。

    確信度はそのランの中の**最大値**を採る (平均ではない)。ランの端は
    隣のトークンへ移る過渡で必ず下がるため、平均を採ると実際より低く出る。
    """
    if log_probs.ndim != 3:
        raise ValueError(f"log_probs must be 3D, got {log_probs.shape}")

    probs = np.exp(log_probs.astype(np.float32, copy=False))
    am = probs.argmax(axis=-1)
    mp = probs.max(axis=-1)
    b, t = am.shape

    results: list[list[FrameToken]] = []
    for i in range(b):
        out: list[FrameToken] = []
        prev = -1
        run_start = 0
        run_max = 0.0
        for j in range(t):
            tok = int(am[i, j])
            conf = float(mp[i, j])
            if tok == prev:
                if tok != blank_id:
                    run_max = max(run_max, conf)
                continue
            # 切り替わり: 前のランを確定
            if prev != -1 and prev != blank_id:
                out.append(FrameToken(
                    token_id=prev, confidence=run_max,
                    frame_start=run_start, frame_end=j - 1,
                ))
            prev = tok
            run_start = j
            run_max = conf
        # 末尾
        if prev != -1 and prev != blank_id:
            out.append(FrameToken(
                token_id=prev, confidence=run_max,
                frame_start=run_start, frame_end=t - 1,
            ))
        results.append(out)
    return results


__all__ = ["FrameToken", "ctc_greedy_decode_frames"]
